src/tree.rs

```diff
@@ -407,28 +407,41 @@ impl Tree {
     /// If both old and new are Some, will modify the value if old is correct.
     /// If Tree is read-only, will do nothing.
     ///
+    /// It will return Unsupported error for the database opened in read-only
+    /// mode. When compare_and_swap fails to update a value it returns a
+    /// CompareAndSwap error with the current value in the tree.
+    ///
     /// # Examples
     ///
     /// ```
     /// let config = sled::ConfigBuilder::new().temporary(true).build();
     /// let t = sled::Db::start(config).unwrap();
     ///
     /// // unique creation
-    /// assert_eq!(t.cas(&[1], None as Option<&[u8]>, Some(&[10])), Ok(Ok(())));
+    /// assert_eq!(
+    ///     t.compare_and_swap(&[1], None as Option<&[u8]>, Some(&[10])),
+    ///     Ok(Ok(()))
+    /// );
     ///
     /// // conditional modification
-    /// assert_eq!(t.cas(&[1], Some(&[10]), Some(&[20])), Ok(Ok(())));
+    /// assert_eq!(
+    ///     t.compare_and_swap(&[1], Some(&[10]), Some(&[20])),
+    ///     Ok(Ok(()))
+    /// );
     ///
     /// // conditional deletion
-    /// assert_eq!(t.cas(&[1], Some(&[20]), None as Option<&[u8]>), Ok(Ok(())));
+    /// assert_eq!(
+    ///     t.compare_and_swap(&[1], Some(&[20]), None as Option<&[u8]>),
+    ///     Ok(Ok(()))
+    /// );
     /// assert_eq!(t.get(&[1]), Ok(None));
     /// ```
-    pub fn cas<K, OV, NV>(
+    pub fn compare_and_swap<K, OV, NV>(
         &self,
         key: K,
         old: Option<OV>,
         new: Option<NV>,
-    ) -> Result<std::result::Result<(), Option<IVec>>>
+    ) -> Result<()>
```

Comment:
This should return a nested result because the results represent different kinds of failures. In the past they were mixed and it led to a lot of bugs. When I switched over to using nested results, the refactor caught several bugs that had not been caught in testing yet as well. That experience led me to believe that very different concerns should not share the same error enums, and it's better to use nesting instead to make the failure handling more clear.

```
) -> Result<std::result::Result<(), CompareAndSwapError>>
```

Reply:
Also feel free to make a public type alias in src/result.rs and re-export it from src/lib.rs:
```
pub type CompareAndSwapResult = std::result::Result<(), CompareAndSwapError>;
```